Documentation/planning/multi-sdk-band-support.md

```diff
@@ -0,0 +1,253 @@
+
+This approach fits naturally into the source-build methodology. Generally, additional SDK bands may be referred to as "subsetted VMRs", since use of a subsetted VMR is not restricted to SDKs only.
+
+### Conceptual splitting of 'previously source-built' artifact types
+
+Today, the term "previously source-built artifacts' refers to those artifacts that were built in a previous product build iteration. Typically N-1. For example, they would refer to the outputs of 8.0.1's build, when building 8.0.2. When building a subsetted VMR, two sets of input sets of source-built artifacts are required. While similar in some ways, they should be kept distinct for clarity. They are referred to as:
+
+- **Previously source-built (PSB)** - Artifacts from a previous servicing/preview iteration, or from a bootstrap build. These may be used when building the current VMR, but may not be bundled into any output.
+- **Current source-built (CSB)** - Artifacts just built, e.g. from a 1xx VMR build. These may be used **and** bundled into the build outputs, since logically they are from the same source-build iteration.
+
+### Supported input artifact combinations and validation thereof
+
+Not all toolsets can be used to build all VMRs. For instance, it is unlikely to be the case that the 1xx SDK will successfully build all 2xx+ bands. Components of newer SDKs frequently take dependencies on newer SDK features. However, the following statements can be made:
+
+- A newer SDK band (e.g. 2xx) may be buildable with an older band.
+- A newer SDK band (e.g. 2xx) must always be able to build itself.
+
+From these statements. we can deduce that the initial release of a newer band (n00) may require a bootstrap, or may be buildable using an older band.
+
+To avoid confusion, Microsoft will:
+
+- Document the set of supported build combinations, including required inputs for all active SDK bands. This provides a guide for distro maintainers.
+- Validate those supported build combinations.
+- Validate 'stage2' builds (that a new SDK can build itself).
+
+## Changes required for the source-build infrastructure
+
+This section details changes to the source build infrastructure to support building of subsetted VMRs.
+
+### Scripting changes
+
+Since conceptually CSB and PSB inputs are different, the scripting should reflect this. Furthermore, the since NuGet and non-NuGet archives are treated similarly within the source-build infrastructure, we should take this opportunity to make changes to the names of existing switches:
+
+- `--with-packages` becomes `--with-previous-artifacts`, reflecting that it represents only previously source built artifacts that may be used in the build, but not redistributed with outputs. When poisoning, this set is poisoned.
+- `--with-current-artifacts` - This switch is added. It points to a directory that contains input NuGet and non-NuGet assets that may be redistributed in the current build. Typically this set would come from the 1xx build.
+
```

Comment:
Do we need an additional argument that allows passing the prebuilts packages tarball?

My understanding it contains packages needed for a successive build to work, which weren't yet buildable from source. So it we should be able to pass it to a build?

Reply:
That's info is currently encoded into the prep script, and so I don't think a new workflow or additional switch would be required.

By the time we have multiple bands (after RTM), we should also never have prebuilts

Reply:
> That's info is currently encoded into the prep script

I have some CI jobs that build a packages dir from a previous build's Artifacts and Prebuilts and then pass that as `--with-packages` to a new build.

It seems I need some way to pass the in Prebuilts when `--with-packages` gets removed?

I assume I can use one of these properties to pass the Artifacts, and the other to pass Prebuilts, and that should work?

Reply:
> I have some CI jobs that build a packages dir from a previous build's Artifacts and Prebuilts and then pass that as --with-packages to a new build.

`--with-previous-artifacts` will be functionally the same as `--with-packages`. So you could use that same workflow. `--with-current-artifacts` would only be used for building newer bands, passing the artifacts built in the 1xx band.

Reply:
> --with-previous-artifacts will be functionally the same as --with-packages. So you could use that same workflow. --with-current-artifacts would only be used for building newer bands, passing the artifacts built in the 1xx band.

If these arguments would accept tarballs (or even tarball urls) it would be handy to not have to extract both tarballs into a single directory to pass it to `--with-previous-artifacts`.

Question: will a build in some way functionally distinguish between packages from `--with-previous-artifacts` vs `--with-current-artifacts`, or will both be 'simply' treated a source of packages (maybe with a preference for `--with-current-artifacts` over `--with-previous-artifacts`?

Reply:
The individual component builds do not distinguish. Just another source. When deciding what versions of a various package to use, current takes precedence over previous. This is the same as today. What just got built (e.g. when building runtime), takes precedence over previously source built.

Reply:
Another way to view this whole effort is that there already is such a concept of "current-artifacts" within the infrastructure today. A shared directory where the outputs of each repo is placed as it is built.

```mermaid
flowchart TD
  previouslysourcebuilt[("previousy source built (--with-packages/prep.sh)")]
  currentsourcebuilt[("current live source built (built in current build invocation)")]
  startbuild["start build"]-.->arcade["Arcade build"]-.->aspnetcore["aspnetcore build"]-.->sdk["sdk build"]-.->preparefinaloutputs["gather source built artifacts"]-.->finishbuid["finish build"]
  previouslysourcebuilt-->arcade
  previouslysourcebuilt-->aspnetcore
  previouslysourcebuilt-->sdk
  currentsourcebuilt-->aspnetcore
  currentsourcebuilt-->sdk
  arcade-->currentsourcebuilt
  aspnetcore-->currentsourcebuilt
  sdk-->currentsourcebuilt
  currentsourcebuilt-->preparefinaloutputs

  subgraph versionselection["version selection"]
    current["current source built version"]-->|overrides|previous["previously source built version"]-->|overrides|repoCheckedInDefault["repo checked in eng/Versions.props"]
  end
```

What this proposal is doing is extending this concept of another set of "current" artifacts that come from another build invocation.

```mermaid
flowchart TD
  previouslysourcebuilt[("previousy source built (--with-packages/prep.sh)")]
  currentlivesourcebuilt[("current source built (built in current build invocation)")]
  currentsourcebuilt[("current source built (--with-current-packages)")]
  startbuild["start build"]-.->roslyn["roslyn build"]-.->sdk["sdk build"]-.->preparefinaloutputs["gather source built artifacts"]-.->finishbuid["finish build"]
  previouslysourcebuilt-->roslyn
  previouslysourcebuilt-->sdk
  currentsourcebuilt-->roslyn
  currentlivesourcebuilt-->sdk
  currentsourcebuilt-->sdk
  roslyn-->currentlivesourcebuilt
  sdk-->currentlivesourcebuilt
  currentlivesourcebuilt-->preparefinaloutputs

  subgraph versionselection["version selection"]
    currentlive["current live source built version"]-->|overrides|current["current source built version"]-->|overrides|previous["previously source built version"]-->|overrides|repoCheckedInDefault["repo checked in eng/Versions.props"]
  end
```

Reply:
Above, the dotted lines represent the execution flow, while solid lines represent the data flow. Hopefully that makes it a little clearer.

Reply:
I added this as an index.